File: src/main/scala/uni/data/BigUtils.scala
package uni.data

import uni.time.*
import scala.util.Try
import scala.math.BigDecimal

//export BigUtils.getMostSpecificType

object BigUtils:
  import uni.data.Big
  import uni.data.Big.Big
  import uni.data.Big.*

  /** A CSV cell value, once typed. Names [[uni.time.UniDateTime]] because that is what
   *  `parseDate` produces -- a union cannot be entered through an implicit conversion, so
   *  a `LocalDateTime` member here would be unreachable from uni's own parser and would
   *  make `toStr` throw `MatchError` on the very values it exists to render. */
  type CVD = UniDateTime|Big|Option[Int]|String|Int

  // ------------------------------------------------------------
  // Core type & sentinels
  // ------------------------------------------------------------

  inline def isBad(n: Big): Boolean = 
    // Cast to AnyRef to enable reference equality check (eq)
    (n.asInstanceOf[AnyRef] eq BigNaN.asInstanceOf[AnyRef]) || (n == BigNaN)

  inline def orBad(opt: Option[Big]): Big =
    opt.getOrElse(BigNaN)

//  val BigZero: Big = Big(0)
  //val BigOne:  Big = Big.one
  //val Hundred: Big = Big.hundred

  private val debug: Boolean =
    Option(System.getenv("DEBUG")).isDefined


  // ------------------------------------------------------------
  // Regex patterns (kept from original)
  // ------------------------------------------------------------

  // The NumPattern1..4 regex set died with the old isNumeric (0.16.0): the
  // patterns existed only to approximate what str2num accepts, and isNumeric
  // now asks str2num directly.

  // ------------------------------------------------------------
  // Character validation
  // ------------------------------------------------------------

  def validNumChar(c: Char): Boolean =
    (c >= '0' && c <= '9') ||
    c == '.' || c == '-' || c == 'E' || c == 'e' ||
    c == '+' || c == '%' || c == '$' || c == ','

  // ------------------------------------------------------------
  // String → Big parsing (BigNaN sentinel preserved)
  // ------------------------------------------------------------

  def str2num(raw: String): Big =
    val trimmed = raw.trim
    if !trimmed.forall(validNumChar) then
      BigNaN
    else
      // `$` and `,` are decoration wherever they appear. The old leading-junk strip
      // (^[^-\.\d]+) discarded ANY non-numeric prefix, so "%50" parsed as 50 -- the
      // percent went as junk and never divided -- while "50%" was 0.5, and "E5"
      // read as 5. Repaired 0.16.0; BigDecimal handles a leading '+' natively.
      val cleaned = trimmed.replaceAll("[$,]", "")

      val normalized =
        if cleaned.startsWith(".") then "0" + cleaned else cleaned

      if normalized.isEmpty then
        BigNaN
      else
        // percent is one trailing suffix, not a character to delete globally --
        // the old replace-all made "5%5" parse as 55
        val nopct = normalized.stripSuffix("%")
        val base: Big = orBad(Try(Big(BigDecimal(nopct))).toOption)
        if isBad(base) then BigNaN
        else if nopct != normalized then base / Big.hundred
        else base

  // ------------------------------------------------------------
  // Numeric detection
  // ------------------------------------------------------------

  /** One definition of "numeric": whatever `str2num` can parse. Delegating keeps
   *  the two functions coherent by construction. The previous pattern-set
   *  answered neither "is this strictly a number" (it accepted `12%`) nor "can
   *  str2num parse this" (it rejected `$1,234.56`): `validNumChar` admitted `$`
   *  and `,` into a strict branch whose patterns could not accept a
   *  currency-with-grouping shape — while the CSV loaders' own type sniffing
   *  stripped the `$` and called the same string numeric. */
  def isNumeric(col: String): Boolean =
    !isBad(str2num(col))

  /** The gate for `getMostSpecificType`, which parses MORE shapes than `str2num`:
   *  a `K`/`M`/`B` scale suffix, a parenthesised negative, a percent. Those were
   *  the other half of what the old `isNumeric` pattern set accepted, and folding
   *  them into `isNumeric` is what made it incoherent — the two callers want
   *  different questions answered, so they get two predicates. Strips the
   *  decorations this function's own body then re-parses, and delegates the
   *  numeric core to `str2num`, so it cannot drift from what follows it. */
  private def isNumericDecorated(col: String): Boolean =
    val s = col.trim
    if s.isEmpty then false
    else
      val unparen =
        if s.startsWith("(") && s.endsWith(")") then s.substring(1, s.length - 1).trim
        else if s.startsWith("(") then s.substring(1).trim // the old set accepted `(  42%`
        else s
      val unsuffixed = unparen.lastOption match
        case Some(c) if "kmbKMB".contains(c) => unparen.dropRight(1)
        case _                               => unparen
      isNumeric(unsuffixed)

  // ------------------------------------------------------------
  // Most specific type: String | Big | DateTime
  // ------------------------------------------------------------

  def getMostSpecificType(raw: String): String | Big | DateTime =
    var col = raw.replaceAll("""[\$]""", "").trim

    if debug then print(s"# rawcol[$raw]\n")

    val value: Any =
      if col.isEmpty then col
      else if isNumericDecorated(col) then
        if debug then print("Numeric match\n")

        var negative = false
        var percent  = false
        var factor   = Big.one

        col = col.replaceAll(",", "")

        if col.contains("%") then
          percent = true
          col = col.replaceAll("%", "")

        def consumeFactor(multiplier: Big): Unit =
          col = col.dropRight(1)
          factor = multiplier

        col.toLowerCase.lastOption.foreach {
          case 'k' => consumeFactor(Big(1_000))
          case 'm' => consumeFactor(Big(1_000_000))
          case 'b' => consumeFactor(Big(1_000_000_000))
          case _   => ()
        }

        if col.startsWith("(") && col.endsWith(")") then
          negative = true
          col = col.substring(1, col.length - 1).trim
        else if col.startsWith("-") then
          negative = true
          col = col.substring(1).trim

        val base = Try(Big(col)).getOrElse(BigNaN)

        val signed =
          if negative && !isBad(base) then -base else base

        val pctAdjusted =
          if percent && !isBad(signed) then signed / Big.hundred else signed

        if isBad(pctAdjusted) then BigNaN
        else pctAdjusted * factor

      // < 6, not < 7 (0.16.0): "1/2/24" is a date. And `parseDate` answers BadDate
      // rather than throwing since the SmartParse migration, so the old
      // Try(..).getOrElse(col) returned the *sentinel* as a DateTime for every
      // unparseable longer string -- the String has to be restored explicitly.
      else if col.length < 6 then col
      else
        val d = parseDate(col)
        if d == BadDate || d == EmptyDate then col else d

    value match
      case bd: BigDecimal => Big(bd)
      case str: String    => str
      case dt: DateTime   => dt
      case other          => other.toString

  // ------------------------------------------------------------
  // Big constructors (explicit, minimal)
  // ------------------------------------------------------------

  /*
  def big(s: String): Big      = apply(s)
  def big(d: Double): Big      = apply(d)
  def big(i: Int): Big         = apply(i)
  def big(l: Long): Big        = apply(l)
  def big(bd: BigDecimal): Big = apply(bd)
  */

  // ------------------------------------------------------------
  // Formatting DSL core
  // ------------------------------------------------------------

  final case class NumFormat(
    colWidth: Int = 9,
    dec: Int = 2,
    factor: Double = 1.0,
    abbreviate: Boolean = false,
    suffix: String = ""
  )

  object NumFormat:
    val Default: NumFormat =
      NumFormat()

    val Abbrev: NumFormat =
      NumFormat(abbreviate = true)

    val Percent: NumFormat =
      NumFormat(dec = 2, factor = 100.0, suffix = "%")

    val IntPercent: NumFormat =
      NumFormat(colWidth = 3, dec = 0, factor = 100.0, suffix = "%")

  // ------------------------------------------------------------
  // Formatting helpers (BigNaN-aware)
  // ------------------------------------------------------------

  def numStr(xx: Big, fmt: NumFormat = NumFormat.Default): String =
    if isBad(xx) then
      " " * (fmt.colWidth - 3) + "N/A"
    else
      val NumFormat(colWidth, dec, factor, abbreviate, suffix) = fmt
      val fmtMain  = s"%${colWidth}.${dec}f"
      val fmtShort = s"%${colWidth - 1}.${dec}f"
      val scaled   = xx * Big(factor)
      // abbreviation is by magnitude (0.16.0): -2.5e9 gets its "B" like +2.5e9 does
      val mag      = scaled.abs

      val raw =
        if abbreviate && mag >= Big(1e9) then
          fmtShort.format((scaled / Big(1e9)).toDouble) + "B"
        else if abbreviate && mag >= Big(1e6) then
          fmtShort.format((scaled / Big(1e6)).toDouble) + "M"
        else
          fmtMain.format(scaled.toDouble)

      // ANY all-zero negative rendering blanks its sign (0.16.0) -- the old exact
      // "-0.00" match let "-0.000" (dec 3) and "-0.00%" (checked after the suffix
      // was appended) keep a minus sign on a zero
      val unsigned =
        if raw.trim.matches("-0(\\.0+)?") then raw.replace("-", " ") else raw
      unsigned + suffix

  def numStrPct(xx: Big, fmt: NumFormat = NumFormat.Percent): String =
    numStr(xx, fmt)

  def num2string(xx: Big, dec: Int = 2, factor: Double = 1.0): String =
    numStr(xx, NumFormat(dec = dec, factor = factor))

  def big2double(xx: Big): Double =
    if isBad(xx) then Double.NaN else xx.toDouble

  def toStr(x: CVD): String = {
    (x: @unchecked) match {
    case s: String        => s
    case n: (Int | Long)  => n.toString
    case BigNaN           => "N/A"
    case b: Big           => b.toString
    case d: UniDateTime   => d.toString("yyyy-MM-dd")
    case Some(oi: Int)    => oi.toString
    case None             => ""
    }
  }

File: rust/src/udata/bigutils.rs
//! Formatting and parsing helpers over [`Big`] — a port of Scala's
//! `uni.data.BigUtils` (exported there as `uni.data.*`).
//!
//! # Fidelity notes
//!
//! [`numStr`] reproduces Java's `String.format("%w.df", double)` exactly: Java rounds
//! the *shortest decimal representation* of the double half-up (its documented `%f`
//! contract), which is also what Rust's `{}` display of an `f64` produces — so the
//! pipeline here is shortest-digits → [`Big::setScale`]`(dec, HalfUp)` → pad. The
//! fixture pins it, including the `-0.00` blanking quirk and the abbreviate branches.
//! (Scala's `format` is locale-dependent; the reference behavior is the `.`-decimal
//! locale the fixtures were generated under.)
//!
//! [`isNumeric`] ports four Scala regexes as character scans. Two authoring accidents
//! in the originals were repaired on the Scala side in 0.16.0 and mirrored here —
//! pattern 3 now takes `K`/`M`/`B` case-insensitively like pattern 1, and pattern 4's
//! decimal point is literal (the unescaped `.` accepted any character, so `12,34E+5`
//! read as numeric). The fixture pins a broad input set so neither side drifts.

#![allow(
    non_snake_case,
    reason = "public items mirror the Scala API name-for-name, so a script kept in both \
              languages needs no mental translation. Internal helpers stay snake_case, \
              so the case says whether a Scala counterpart exists."
)]

use crate::udata::Big;
use crate::udata::big::RoundingMode;

/// The characters `str2num`/`isNumeric` treat as potentially numeric.
fn valid_num_char(c: char) -> bool {
    c.is_ascii_digit() || matches!(c, '.' | '-' | 'E' | 'e' | '+' | '%' | '$' | ',')
}

/// Is `n` the BigNaN sentinel?
#[must_use]
pub fn isBad(n: &Big) -> bool {
    n.isNaN()
}

/// The value, or BigNaN for `None`.
#[must_use]
pub fn orBad(opt: Option<Big>) -> Big {
    opt.unwrap_or_else(Big::nan)
}

/// Lenient string → [`Big`]: `$`/`,` are decoration wherever they appear, one
/// trailing `%` divides by 100, and anything else answers BigNaN. The 0.16.0 Scala,
/// step for step (the old leading-junk strip made `"%50"` parse as 50, and the old
/// global %-removal made `"5%5"` parse as 55).
#[must_use]
pub fn str2num(raw: &str) -> Big {
    let trimmed = raw.trim();
    if !trimmed.chars().all(valid_num_char) {
        return Big::nan();
    }
    let cleaned: String = trimmed.chars().filter(|&c| c != '$' && c != ',').collect();
    let normalized = if cleaned.starts_with('.') {
        format!("0{cleaned}")
    } else {
        cleaned
    };
    if normalized.is_empty() {
        return Big::nan();
    }
    let nopct = normalized.strip_suffix('%').unwrap_or(&normalized);
    let base = Big::parse(nopct);
    if base.isNaN() {
        Big::nan()
    } else if nopct != normalized {
        base.div(&Big::parse("100"))
    } else {
        base
    }
}

// ── The four NumPattern matchers, as character scans ─────────────────────────
//
// Each is a full-string match, like Scala's `s match { case NumPattern(..) }`.

/// One definition of "numeric": whatever `str2num` can parse. Delegating keeps
/// the two functions coherent by construction — the previous pattern-set
/// rejected `$1,234.56` while `str2num` (and the CSV loaders' type sniffing)
/// parsed it happily. Mirrors the Scala.
pub fn isNumeric(col: &str) -> bool {
    !isBad(&str2num(col))
}

// ── Formatting ────────────────────────────────────────────────────────────────

/// Scala's `NumFormat` formatting descriptor.
#[derive(Debug, Clone)]
pub struct NumFormat {
    pub colWidth: i32,
    pub dec: i32,
    pub factor: f64,
    pub abbreviate: bool,
    pub suffix: String,
}

impl Default for NumFormat {
    fn default() -> Self {
        Self {
            colWidth: 9,
            dec: 2,
            factor: 1.0,
            abbreviate: false,
            suffix: String::new(),
        }
    }
}

impl NumFormat {
    /// 9-wide, 2 decimals.
    #[must_use]
    pub fn Default() -> Self {
        Self::default()
    }

    /// Billions/millions collapse to `B`/`M`.
    #[must_use]
    pub fn Abbrev() -> Self {
        Self {
            abbreviate: true,
            ..Self::default()
        }
    }

    /// ×100 with a `%` suffix.
    #[must_use]
    pub fn Percent() -> Self {
        Self {
            factor: 100.0,
            suffix: "%".to_string(),
            ..Self::default()
        }
    }

    /// 3-wide integer percent.
    #[must_use]
    pub fn IntPercent() -> Self {
        Self {
            colWidth: 3,
            dec: 0,
            factor: 100.0,
            suffix: "%".to_string(),
            ..Self::default()
        }
    }
}

/// Java's `String.format("%<width>.<dec>f", v)`: shortest decimal digits of the
/// double, rounded half-up at `dec` decimals, right-justified to `width`.
///
/// Public because a script ported from Scala needs it directly: `f"$x%.6f"` there
/// rounds the *shortest decimal representation* half-up, while Rust's `{:.6}` rounds
/// the *exact binary value* half-to-even. Those disagree whenever the two fall on
/// opposite sides of the boundary — rare per value, but a byte-identical demo pair
/// printing thousands of numbers cannot rely on never hitting one.
///
/// Stays snake_case under the naming contract: it has no `uni` Scala counterpart to
/// mirror, because on that side the behaviour comes from the `f` interpolator, a
/// language feature rather than a library method.
pub fn java_format_f(v: f64, width: i32, dec: i32) -> String {
    let body = if v.is_nan() {
        "NaN".to_string()
    } else if v.is_infinite() {
        if v < 0.0 {
            "-Infinity".to_string()
        } else {
            "Infinity".to_string()
        }
    } else {
        // The formatter keeps the double's sign even when the digits round to zero.
        let neg = v < 0.0 || (v == 0.0 && v.is_sign_negative());
        let digits = format!("{}", v.abs()); // shortest round-trip, positional
        let rounded = Big::parse(&digits)
            .setScale(dec, RoundingMode::HalfUp)
            .toPlainString();
        if neg { format!("-{rounded}") } else { rounded }
    };
    let pad = width.max(0) as usize;
    if body.len() < pad {
        format!("{}{}", " ".repeat(pad - body.len()), body)
    } else {
        body
    }
}

/// BigNaN-aware column formatting: `N/A` for the sentinel, otherwise
/// `%colWidth.decf` of `xx * factor` (with `B`/`M` abbreviation when asked), the
/// suffix appended, and an exact `-0.00` blanked to ` 0.00` — the Scala, verbatim.
#[must_use]
pub fn numStr(xx: &Big, fmt: &NumFormat) -> String {
    if isBad(xx) {
        return format!("{}N/A", " ".repeat((fmt.colWidth - 3).max(0) as usize));
    }
    let scaled = xx.mul(&Big::from_f64(fmt.factor));
    // abbreviation is by magnitude (0.16.0): -2.5e9 gets its "B" like +2.5e9 does
    let mag = scaled.abs();
    let raw = if fmt.abbreviate && mag.compare(&Big::from_f64(1e9)) >= 0 {
        let v = scaled.div(&Big::from_f64(1e9)).toDouble();
        format!("{}B", java_format_f(v, fmt.colWidth - 1, fmt.dec))
    } else if fmt.abbreviate && mag.compare(&Big::from_f64(1e6)) >= 0 {
        let v = scaled.div(&Big::from_f64(1e6)).toDouble();
        format!("{}M", java_format_f(v, fmt.colWidth - 1, fmt.dec))
    } else {
        java_format_f(scaled.toDouble(), fmt.colWidth, fmt.dec)
    };
    // ANY all-zero negative rendering blanks its sign, before the suffix (0.16.0):
    // "-0.000" at dec 3 and "-0.00%" no longer keep a minus on a zero
    let unsigned = if is_negative_zero_render(raw.trim()) {
        raw.replace('-', " ")
    } else {
        raw
    };
    format!("{unsigned}{}", fmt.suffix)
}

/// `-0`, `-0.0`, `-0.00`, … — the shapes whose minus sign [`numStr`] blanks.
fn is_negative_zero_render(t: &str) -> bool {
    t == "-0" || (t.len() > 3 && t.starts_with("-0.") && t[3..].bytes().all(|b| b == b'0'))
}

/// [`numStr`] with the percent format.
#[must_use]
pub fn numStrPct(xx: &Big, fmt: &NumFormat) -> String {
    numStr(xx, fmt)
}

/// [`numStr`] with just decimals and a factor.
#[must_use]
pub fn num2string(xx: &Big, dec: i32, factor: f64) -> String {
    numStr(
        xx,
        &NumFormat {
            dec,
            factor,
            ..NumFormat::default()
        },
    )
}

/// The double, or `NaN` for the sentinel.
#[must_use]
pub fn big2double(xx: &Big) -> f64 {
    if isBad(xx) { f64::NAN } else { xx.toDouble() }
}
